### ISSUE TITLE
Bugs and Wishlist for Intellij Flutter support...

### ISSUE THREAD

**Issue by patniemeyer** [flutter/flutter-intellij#4165]
**Bugs(?)**
- Renaming a file should update import statements throughout the code.
- Overriding initState() in flutter should generate the default call to super.initState() (since intellj warns about this)
- Generated object instantiations in Dart no longer need the “new” keyword. (e.g. refactor->extract widget)
- Why does command-enter to “force complete” a statement add a newline in dart/flutter? (Is this configurable?)

**Wishlist**
- Ability to extract a class or widget to its own file.  It’s really common to create a bunch of small widgets in a file and then as they grow want to break them out into their own files. 
- Refactoring should allow removal of a “builder” style widget wrapper.
- Option when generating constructors to use the named arguments style (which would simply wrap the arguments list in {} curly braces, saving you a step if you do that every time).
- Option to preserve single extra lines at the start of a class, array, etc.

## Version info

Flutter version 1.12.15
IntelliJ IDEA 2019.3 (Ultimate Edition), Build #IU-193.5233.102, built on November 27, 2019
Latest version of all plugins.

**Comment by devoncarew** [comment]
Thanks for all the feedback! We'll likely need to split these into distinct issues, and track them in a few different repos (dart-lang/sdk, here, and the Dart IntelliJ plugin). Comments in-line.

cc @scheglov @pq @bwilkerson 

> **Bugs(?)**
> 
> * Renaming a file should update import statements throughout the code.
> * Overriding initState() in flutter should generate the default call to super.initState() (since intellj warns about this)

Agreed

> * Generated object instantiations in Dart no longer need the “new” keyword. (e.g. refactor->extract widget)
> * Why does command-enter to “force complete” a statement add a newline in dart/flutter? (Is this configurable?)

I'm not sure; cc @jwren.

> 
> **Wishlist**
> 
> * Ability to extract a class or widget to its own file.  It’s really common to create a bunch of small widgets in a file and then as they grow want to break them out into their own files.
> * Refactoring should allow removal of a “builder” style widget wrapper.

Can you give an example?

> * Option when generating constructors to use the named arguments style (which would simply wrap the arguments list in {} curly braces, saving you a step if you do that every time).
> * Option to preserve single extra lines at the start of a class, array, etc.

This is when formatting? When performing a refactor?

**Comment by patniemeyer** [comment]
Thanks for your response and sorry for dumping these all in one report!

An example of "Refactoring should allow removal of a “builder” style widget wrapper" -

Perhaps I should say "context options" instead of "refactoring".   When working in a widget tree Intellij offers context options including "wrap with widget" or "remove widget".  Intellij currently supports adding a "stream builder" type widget but cannot subsequently remove that widget or other "builder" style widgets.  I can see the difficult cases involved, but wish for it nonetheless :)

Re: "Option to preserve single extra lines at the start of a class, array, etc."

Now that I look more closely I think this may be an issue with `dartfmt` and not Intellij.  Perhaps the wishlist item should be for more control over or a replacement for that formatting.

Thanks again!

**Comment by jwren** [comment]
By "force complete" I assume you are referencing statement completion?  In general we try to match other languages, if there is a mis-match we should create a bug on the Dart Plugin. @alexander-doroshko

**Comment by patniemeyer** [comment]
Yes, thanks, I was referring to "complete statement".  If I type, e.g.:
`String foo = "foo"`
and then hit the complete statement shortcut it will add the semicolon and then *also* immediately start a new line.  This is often not appropriate and Intellilj does not do this when editing e.g. Java or Typescript.  The effect is also that in vim you are left in editing mode, which is also different behavior from Java and Typescript editing.

**Comment by alexander-doroshko** [comment]
@jwren Statement Completion is handled by the Dart Analysis Server (implemented by @stevemessick)

**Comment by pq** [comment]
> * Generated object instantiations in Dart no longer need the “new” keyword. (e.g. refactor-> extract widget)

This one should be complete as of https://github.com/dart-lang/sdk/commit/082e82ed5e5885f9f1cdc86e771b7b265089040f.

> * Why does command-enter to “force complete” a statement add a newline in dart/flutter? (Is this configurable?)

Could you give a concrete example?  (Looking at test cases now.)

Thanks!
